### ISSUE TITLE
No graceful shutdown when pod gets evicted

### ISSUE THREAD

**Issue by SebastianStehle** [kubernetes/kubernetes#71319]
**What happened**:

I am not sure if it is really a bug.

Yesterday the eviction manager cleaned a pod to free up memory. But the graceful shutdown process failed and the container has been terminated.

    I1120 21:46:00.617083 1300 kuberuntime_container.go:582] Killing container "docker://7b9af6ed59606201c15daf1cec2e263415b525873ab58415b562e4e2de187949" with 30 second grace period

When looking at the log, there is no further output from the this file, which is strange, because it should log whether the shutdown has been successful or not: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kuberuntime/kuberuntime_container.go#L581

A second later the pod gets killed:

    21:46:01.712812 1300 kubelet_pods.go:1118] Killing unwanted pod "squidex-854d46775f-kd8wn"

The full logs are https://www.dropbox.com/s/5q8d0gjp81yxkoj/gce_instance__logs__2018-11-20T10-47.json?dl=0

**What you expected to happen**:

* A graceful shutdown.

**How to reproduce it (as minimally and precisely as possible)**:

* I don't know. I tried to use `kubectl drain` to reproduce it but it does not work.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): v1.9.7-gke.11
- Cloud provider or hardware configuration: GoogleCloud
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

/kind bug

**Comment by tahoward** [comment]
I'm seeing this issue as well. Creating an eviction resource from k8s API terminates the pods regardless of how the container handles SIGTERM event or if it has a preStop CMD.

**Comment by tedyu** [comment]
@SebastianStehle
The full log is no longer accessible.
If you still have it, can you repost ?

thanks

**Comment by tedyu** [comment]
From kubelet_pods.go :
```
				klog.V(2).Infof("Killing unwanted pod %q", runningPod.Name)
				err := kl.killPod(apiPod, runningPod, nil, nil)
```
It seems the 'Killing unwanted pod' log should have appeared prior to the 'Killing container' log (killPod ends up calling killContainersWithSyncResult).

**Comment by tedyu** [comment]
I wonder if the following change would respect the grace period:
```
diff --git a/pkg/kubelet/kuberuntime/kuberuntime_manager.go b/pkg/kubelet/kuberuntime/kuberuntime_manager.go
index 0a5e666da6..c7e4419563 100644
--- a/pkg/kubelet/kuberuntime/kuberuntime_manager.go
+++ b/pkg/kubelet/kuberuntime/kuberuntime_manager.go
@@ -839,9 +839,16 @@ func (m *kubeGenericRuntimeManager) KillPod(pod *v1.Pod, runningPod kubecontaine
 // Note: The pod passed in could be *nil* when kubelet restarted.
 func (m *kubeGenericRuntimeManager) killPodWithSyncResult(pod *v1.Pod, runningPod kubecontainer.Pod, gracePeriodOverride *int64) (result kubecontainer.PodSyncResult) {
        killContainerResults := m.killContainersWithSyncResult(pod, runningPod, gracePeriodOverride)
+       seenErr := false
        for _, containerResult := range killContainerResults {
+               if containerResult.Error != nil {
+                       seenErr = true
+               }
                result.AddSyncResult(containerResult)
        }
+       if !seenErr {
+               time.Sleep(time.Second * time.Duration(*gracePeriodOverride))
+       }

        // stop sandbox, the sandbox will be removed in GarbageCollect
        killSandboxResult := kubecontainer.NewSyncResult(kubecontainer.KillPodSandbox, runningPod.ID)
```

**Comment by aka-bo** [comment]
I ran into this issue as well and was able to trace back to #72717 which was fixed in #72730. Essentially, the evictions API would fail to honor graceful terminations when called with a `nil` deleteOptions. 

I was able to workaround this issue by passing an empty dict instead:

```
{
    "apiVersion": "policy/v1beta1",
    "kind": "Eviction",
    "metadata": {
        "name": "my-pod",
        "namespace": "default"
    },
    "deleteOptions": {}
}
```
